### ISSUE TITLE
Importing arbitrary models

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#349]
Although @jhmatthews and I thought we had imported models working they do not for the case where one cannot define windcones to bound the wind.  The underlying problem is that we generally regard a cell as being in or out of the wind, and we use windcones basically to exclude cells which are not in the wind.  This is not the case for the FU Ori models where the wind region is curved, and therefore we have to deal with windcells that are not in the wind.  The problem is tricky in part because we treat grid cells which are not in the wind quite differently from those that are in the wind.  For example, we do not create a plasma pointer for them.   This work is currently being done on the import branch.  

The state of play as of the beginning of today was that it appears that some photons that should have gone through "good" wind cells, did not, for a variety of input files.  One of which is 
[today.pf.txt](https://github.com/agnwinds/python/files/1779053/today.pf.txt)

What James and I observed for a file like this was that there appeared to be more photons just outside wind_rmax than there were inside wind_rmax when you set wind_rmax less than the size of the grid.

The reason for this was as follows:

Typically photons move in space first to the inner radius.  At that point, with an imported model.   Either the first cell they encounter after this, will be in the wind or it will not.

If the cell is in the wind, then the photon progresses normally, 

If this cell is not in the wind however, Python says it is not in the wind, and if it is not in the wind, the photon is transported to the next boundary, which typically is the outer boundary

Both type of photons, find themselves just outside the outer boundary, and if these are in the wind they started getting transported in the wind again.

So you get more photons outside the boundary than inside.  They progress until the run out of the wind cells in the wind, at which point they go to infinity.

To fix this, we need to track photons in these wind cells, but prevent photons from doing anything else.  Fortunately, we do have some current concept that there are some cells that are inside the wind cones but still in the wind, which have the label W_IGNORE.  These were supposed to be cells with too little volume to be concerned about.

So the current version of the import_cylind   changes all of the wind cells labelled NOT_INWIND to W_IGNORE.  This does help some, but we are still seeing more photons at the outer boundary, though the problem is not as great as previously.  However, there is a new problem which we see which we are getting trapped photons, which prevent the program from completing (including with the Kelly's initial model), so the reason for this has to be tracked down. 

I tried trapping these trapped photons, and giving them a "push" but that has not worked thus far.

Note to James:  There is a new diagnostic that one can use to save_photons on the import branch.  When used it prints the postions of photons as they go through the grid, or a goodly number of them anyway.

**Comment by kslong** [comment]
The current state of play here is that we originally succeeded in importing models in multiple coordinate systems where they were models that were generated from python originally, but that software failed to work on models from Lee Hartmann, where we needed to do a better job in dealing with the situation where the edges of the wind were not determined by wind cones.  We then modified python to handle a cylindrical import properly, but we have not made coded or tested the same fixes for spherical or rtheta grids.  Currently the program thows and error (and quits) if you try to import anything but a cylindrical grid.

**Comment by kslong** [comment]
The status of the issue at the time of the Belfast meeting remains the same.  

- We are able to import Hartmann models in cylindrical coordinates, but we are not able to import a generic model in either spherical or r-theta coordinates.  
- The challenge is that in an arbitrary model the boundaries of the wind are not simple cones which is the way our parametric models and also the way Nick's hydro models work.  Therefor the code needs to look in the individual cells as photons are passing through it to determine whether it is in the wind or not.  One still needs to worry about boundaries like rmax, etc.
- There is a python script that can convert one of the outputs of windsave2table into something that is appropriate to import.  These are call import_1d.py, etc.
- Before we discovered the problem associated with finding which cells were in the wind we had a versions of the code which could import spherical and rtheta models of the form built by the python script mentioned above.  Currently, the program throws an error and quits if you try to import one of these models and quits, but the code is still there.
- At the time we address this problem, we should consider incorporating the ability to read in zeus/hydro models with this generic software.

**Comment by kslong** [comment]
There is more information about the history of this in #309

**Comment by kslong** [comment]
I have merged changes into dev that now allow one to import polar (rtheta) and spherical models.  One can create test models from normal python runs these using import_rtheta.py and import_1d.py  The changes should not affect any other capabilities of python.  It turned out that very few additional changes were needed from the code we created to accept cylindrical models, @jhmatthews.  I am closing the issue, but folks should still use this capability with care (particularly in the case where the  wind is shaped in an unusal way.  Note that this  code does not make use windcones (either for the cylindrical models we have tested previsouly, or for the rtheta ones that are not incorporated into the code)

In principle @saultyevil (and @lazygun37) one could do some testing of optically thick 1d models by with this import capability (by creating a run of densities and a source with your Eddington code, and adapting this so it can be accepted by Python) but one would need to set the inner radius to something large compared to the physical thickness of the region to avoid differences between the 1d linear and 1d spherical code.
